kernel/sched.c

```diff
@@ -1123,12 +1123,13 @@ int32_t z_impl_k_sleep(k_timeout_t timeout)
 
 	ticks = z_tick_sleep(ticks);
 
-	int32_t ret = K_TIMEOUT_EQ(timeout, K_FOREVER) ? K_TICKS_FOREVER :
+	/* k_sleep() still returns 32 bit milliseconds for compatibility */
+	int64_t ms = K_TIMEOUT_EQ(timeout, K_FOREVER) ? K_TICKS_FOREVER :
 		      k_ticks_to_ms_ceil64(ticks);
+	ms = CLAMP(ms, 0, INT_MAX);
```

Comment:
As is, this unfortunately results in a behavior inconsistency when waking up the edge case of a k_sleep(K_FOREVER). When 64-bit timeouts are used, we will get 0, and when 32-bit timeouts are used we will get INT_MAX.

Reply:
Oops, good catch.  Precedence bug, must clamp before the short circuit, not after, because of the signedness mismatch.  tests/kernel/sleep caught it too.